### PR TITLE
feat: Adds CardExpirationDateElement

### PR DESCRIPTION
**Pull request by dhudec**
## Description

- Adds `CardExpirationDateElement`
- Adds support for referencing arbitrary values exposed by an element

## Testing required outside of automated testing?

- [x] Not Applicable

### Screenshots (if appropriate):

image

## Rollback / Rollforward Procedure

- [x] Roll Forward
- [ ] Roll Back

## Reviewer Checklist

- [ ] Description of Change
- [ ] Description of outside testing if applicable.
- [ ] Description of Roll Forward / Backward Procedure
- [ ] Documentation updated for Change

### CODE REVIEW

File: lib/src/main/java/com/basistheory/android/view/CardExpirationDateElement.kt [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    private fun getMonthValue(): String? =
+        getText()
+            ?.split("/")
+            ?.elementAtOrNull(0)
+
+    private fun getYearValue(): String? =
+        getText()
+            ?.split("/")
+            ?.elementAtOrNull(1)
```
Comment (dhudec) [inline lib/src/main/java/com/basistheory/android/view/CardExpirationDateElement.kt:46]: Incorrect! I need to prepend "20" to the year or else this won't be usable to create a card token

**Review by kevinperaza — APPROVED** [review]
Good stuff 🚀

File: lib/build.gradle [hunk 2]
```diff
@@ -68,6 +68,7 @@ dependencies {
     implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-core:1.6.4'
     implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-android:1.6.4'
     implementation 'org.apache.commons:commons-lang3:3.12.0'
+    implementation 'org.threeten:threetenbp:1.6.4'
```
Comment (jleon15) [inline lib/build.gradle:71]: what is this for?

Comment (jleon15) [reply lib/build.gradle:71]: oo is it cuz of the java.time issue you told me?

File: lib/src/main/java/com/basistheory/android/view/mask/MaskWatcher.kt [hunk 3]
```diff
@@ -1,47 +0,0 @@
-package com.basistheory.android.view.mask
```
Comment (jleon15) [inline lib/src/main/java/com/basistheory/android/view/mask/MaskWatcher.kt:1]: ok this PR is going to give me lots of conflicts :rip:

File: lib/src/main/java/com/basistheory/android/view/mask/MaskResult.kt [hunk 4]
```diff
@@ -1,17 +0,0 @@
-package com.basistheory.android.view.mask
-
-import android.text.Editable
-
-internal data class MaskResult(
-    val maskedValue: String,
-    val unMaskedValue: String,
```
Comment (jleon15) [inline lib/src/main/java/com/basistheory/android/view/mask/MaskResult.kt:7]: i guess the unmasked value is not needed cuz we already have it on the element?

File: lib/src/main/java/com/basistheory/android/view/TextElement.kt [hunk 5]
```diff
@@ -16,15 +16,17 @@ import com.basistheory.android.event.BlurEvent
...
+import com.basistheory.android.view.mask.Mask
 
 open class TextElement : FrameLayout {
     private var attrs: AttributeSet? = null
     private var defStyleAttr: Int = androidx.appcompat.R.attr.editTextStyle
     private var input: AppCompatEditText = AppCompatEditText(context, attrs, defStyleAttr)
     private var defaultBackground = input.background // todo: find a better way to reference this
     private val eventListeners = ElementEventListeners()
-    private var maskWatcher: MaskWatcher? = null
+    private var maskValue: Mask? = null
```
Comment (jleon15) [inline lib/src/main/java/com/basistheory/android/view/TextElement.kt:29]: I wouldn't call this maskValue, I even think the name of the class itself is off because it is more of like a mask helper or enforcer

Comment (jleon15) [reply lib/src/main/java/com/basistheory/android/view/TextElement.kt:29]: or masker lol

File: lib/src/main/java/com/basistheory/android/view/TextElement.kt [hunk 6]
```diff
@@ -135,37 +139,70 @@ open class TextElement : FrameLayout {
...
-    private fun addMask(mask: List<Any>) {
-        maskWatcher = MaskWatcher(mask)
-        input.addTextChangedListener(maskWatcher)
-    }
-
-    private fun removeMask() {
-        input.removeTextChangedListener(maskWatcher)
-        maskWatcher = null
-    }
+    protected open fun transformUserInput(userInput: String?): String? = userInput
```
Comment (jleon15) [inline lib/src/main/java/com/basistheory/android/view/TextElement.kt:142]: why is this needed if we have transform above?

File: lib/src/main/java/com/basistheory/android/view/validation/FutureDateValidator.kt [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+fun futureDateValidator(value: String?, currentDate: LocalDate = LocalDate.now()): Boolean {
+    try {
+        if (value.isNullOrEmpty()) return false
+
+        val segments = value.split("/")
+        if (segments.count() != 2) return false
+
+        val month = segments[0].toInt()
+        val year = 2000 + segments[1].toInt()
```
Comment (jleon15) [inline lib/src/main/java/com/basistheory/android/view/validation/FutureDateValidator.kt:13]: and in 78 years?

File: lib/build.gradle (hunk 2 above)
Comment (dhudec) [reply lib/build.gradle:71]: Yes, that was previously only used from the example app before, but expiration date validation required it also in the lib module.

File: lib/src/main/java/com/basistheory/android/view/mask/MaskResult.kt (hunk 4 above)
Comment (dhudec) [reply lib/src/main/java/com/basistheory/android/view/mask/MaskResult.kt:7]: Correct, and it was never used

File: lib/src/main/java/com/basistheory/android/view/TextElement.kt (hunk 6 above)
Comment (dhudec) [reply lib/src/main/java/com/basistheory/android/view/TextElement.kt:142]: This is an internal transform that an element can use to transform user input. The other transform is publicly exposed and is used to transform before tokenization - I can't change that name so that it matches the web/ios interfaces.

File: lib/src/main/java/com/basistheory/android/view/validation/FutureDateValidator.kt (hunk 7 above)
Comment (dhudec) [reply lib/src/main/java/com/basistheory/android/view/validation/FutureDateValidator.kt:13]: 💥

File: lib/src/main/java/com/basistheory/android/view/TextElement.kt (hunk 5 above)
Comment (dhudec) [reply lib/src/main/java/com/basistheory/android/view/TextElement.kt:29]: I can't think of a better name ATM, maybe we can change it with another PR
